File: pythonAlgorithms/Hamming.py
class Hamming:
    def __init__(self, n: int, m: int, code: str) -> None:
        self.r = n - m
        self.m = m
        self.n = n

        if not ((m + self.r + 1) <= 2**self.r):
            raise Exception("\n\nla secuencia no cumple la condicion para ser valida: (m + r + 1) <= 2 ** r\n\n")
        else:
            self.code = code


    def get_parities(self):
        self.r = 0
        self.m = len(self.code)

        while (2 ** self.r < self.m + self.r + 1):
            self.r += 1
        
        self.final_len = self.m + self.r
        return self.r


    def get_by_index(self, index: int, secuence: str):
        val = 0

        for i in range(1, self.final_len + 1):
            if i & index != 0:
                if secuence[i].startswith("p"):
                    continue
                val ^= int(secuence[i])

        return val   


    def set_parity_positions(self):
        result = [None for _ in range(self.final_len+1)]
        j = 0

        for i in range(1, self.final_len+1):
            if i == 2**j:
                result[i] = "p" + str(j+1)
                j += 1
            else:
                result[i] = self.code[i-j-1]
            
        for i in range(self.r):
            index = 2**i
            val = self.get_by_index(index=index, secuence=result)
            result[index] = f"({str(val)})"
        
        return result
    

    def encode(self, arr: list[str]):
        encoded = ""
        insides = ""
        for i in range(1, len(arr)):
            if arr[i] is not None:
                encoded += arr[i]
                if arr[i].endswith(")") and arr[i].startswith("("):
                    insides += arr[i][1:-1]

        return (encoded.replace("(", "").replace(")", ""), insides)


    def hamming(self):
        self.get_parities()
        hamming_arr = self.set_parity_positions()
        self.hamming_encoded, self.parities = self.encode(hamming_arr)
    

    def print_results(self):
        print(f"""Hamming:
            Secuencia ingresada: {self.code}
            Cantidad de bits de paridad: {self.get_parities()}
            Codificacion: {self.hamming_encoded}
            Bits de paridad: {self.parities}
        """)

File: javaAlgorithms/Hamming.java
public class Hamming {
    private int r;
    private int m;
    private int n;
    private String code;
    private int finalLen;
    private String hammingEncoded;
    public String parities;

    public Hamming(int n, int m, String code) {
        this.r = n - m;
        this.m = m;
        this.n = n;

        if (!((m + this.r + 1) <= Math.pow(2, this.r))) {
            throw new RuntimeException("\n\nla secuencia no cumple la condicion para ser valida: (m + r + 1) <= 2 ** r\n\n");
        } else {
            this.code = code;
        }
    }

    public int getParities() {
        this.r = 0;
        this.m = this.code.length();

        while (Math.pow(2, this.r) < this.m + this.r + 1) {
            this.r++;
        }

        this.finalLen = this.m + this.r;
        return this.r;
    }

    public int getByIndex(int index, String sequence) {
        int val = 0;

        for (int i = 1; i <= this.finalLen; i++) {
            if ((i & index) != 0) {
                if (sequence.charAt(i) == 'p') {
                    continue;
                }
                val ^= Character.getNumericValue(sequence.charAt(i));
            }
        }

        return val;
    }

    public String[] setParityPositions() {
        String[] result = new String[this.finalLen + 1];
        int j = 0;

        for (int i = 1; i <= this.finalLen; i++) {
            if (i == Math.pow(2, j)) {
                result[i] = "p" + (j + 1);
                j++;
            } else {
                result[i] = String.valueOf(this.code.charAt(i - j - 1));
            }
        }

        for (int i = 0; i < this.r; i++) {
            int index = (int) Math.pow(2, i);
            int val = getByIndex(index, String.join("", result));
            result[index] = "(" + val + ")";
        }

        return result;
    }

    public String[] encode(String[] arr) {
        String encoded = "";
        String insides = "";

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] != null) {
                encoded += arr[i];
                if (arr[i].endsWith(")") && arr[i].startsWith("(")) {
                    insides += arr[i].substring(1, arr[i].length() - 1);
                }
            }
        }

        return new String[]{encoded.replace("(", "").replace(")", ""), insides};
    }

    public void hamming() {
        this.getParities();
        String[] hammingArr = this.setParityPositions();
        String[] result = this.encode(hammingArr);
        this.hammingEncoded = result[0];
        this.parities = result[1];
    }

    public void printResults() {
        System.out.println("Hamming:\n" +
                "Secuencia ingresada: " + this.code + "\n" +
                "Cantidad de bits de paridad: " + this.getParities() + "\n" +
                "Codificacion: " + this.hammingEncoded + "\n" +
                "Bits de paridad: " + this.parities + "\n");
    }
}
